platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,241 @@
+to reduce its reservation count to 0, the object will be destroyed, and its reservation count will be set to -1.
+
+---
+
+We rely on this mechanism to manage the number of states kept in memory.
+
+**Important**: It is the responsibility of the client code to ensure that the state is released once it is no longer needed.
+
+Once the reservation count reaches 0, the state becomes eligible for garbage collection.
+
+---
+
+### Java Classes
+
+`com.swirlds.state.State` provides the following lifecycle-related methods:
+
+- `State copy()`: Creates a mutable copy of the state. The source state becomes immutable.  
+- `void computeHash()`: Computes the hash of the immutable state.  
+- `void createSnapshot(Path targetPath)`: Creates a snapshot of the state at the specified path.  
+- `State loadSnapshot(Path sourcePath)`: Loads a snapshot from the specified path.  
+
+These methods correspond to the state lifecycle as follows:
+
+[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)
+
+---
+
+### `SwirldStateManager`
+
+The `SwirldStateManager` class handles all interactions with the state object in the context of the platform.  
+
+The current implementation of `SwirldStateManager` has two distinct responsibilities:
+
+1. **State Management**:  
+   - Managing references to the latest immutable state and the latest mutable state.  
+   - Creating copies of the state and updating references.  
+   This functionality belongs in `swirlds-state-api`.  
+
+2. **Platform-Related Event Handling** (belongs to `swirlds-platform-core`):  
+   - Handling consensus rounds  
+   - Sealing consensus rounds  
+   - Keeping the `freeze time` parameter up to date  
+   - Providing the state for signing  
+   - Initializing the state by extracting it from the signed state  
+
+## Proposed Changes
+
+All lifecycle-related methods from `com.swirlds.state.State` should be moved to a separate class responsible for managing the state lifecycle. This will provide a clearer separation of concerns.  
+
+- `com.swirlds.state.State` should focus solely on accessing and modifying the state.  
+- The state lifecycle will be managed externally.
+
+The `SwirldStateManager` class requires refactoring to separate its two distinct responsibilities. `swirlds-state-impl` should not include any platform-specific details but should provide the necessary mechanisms for the platform to interact with the state and its lifecycle.
+
+### `StateLifecycleManager` Interface
+
+To achieve this, `swirlds-state-api` will define a `StateLifecycleManager` interface, with an implementation `StateLifecycleManagerImpl` in `swirlds-state-impl`. The interface will handle:
+
+- Creating/loading an initial state. If no snapshot is available, a new state will be created.  
+- Managing references to the latest immutable state and the latest mutable state.
+- Resetting the mutable state to the latest immutable state.
```

Comment:
I don't understand this resetting functionality, can you add a bit more detail please? During a reset, what happens to the mutable state and to the immutable state?

Reply:
> During a reset, what happens to the mutable state and to the immutable state?

The previous mutable state get discarded from the `StateLifecycleManager` (it's a client's code responsibility to release and destroy it). Current immutable state remains unchanged.

Reply:
Exactly.  Block Node will need the ability to reset to the last immutable state and begin applying changes again from that point.
Examples for when this might be needed include
1. A block fails verification, so we have to request it be resent or request it from another source.
1. A stream is interrupted mid-block (e.g. by a network fault) and we need to request resend or from another source.